executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,430 @@
+use core::fmt;
+use std::{collections::BTreeMap, ops::RangeFrom};
+
+use powdr_ast::{
+    analyzed::{
+        AlgebraicExpression, AlgebraicUnaryOperator, ConnectIdentity, Identity as AnalyzedIdentity,
+        LookupIdentity, PermutationIdentity, PhantomLookupIdentity, PhantomPermutationIdentity,
+        PolynomialIdentity, SelectedExpressions,
+    },
+    parsed::visitor::Children,
+};
+use powdr_number::FieldElement;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct BusSend<T> {
+    /// The ID is globally unique among identities.
+    pub id: u64,
+    /// The ID of the bus interaction, used to find a matching receive.
+    pub interaction_id: AlgebraicExpression<T>,
+    /// The tuple sent to the bus, with a selector.
+    pub selected_tuple: SelectedExpressions<T>,
+}
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct BusReceive<T> {
+    /// The ID is globally unique among identities.
+    pub id: u64,
+    /// The ID of the bus interaction, to be matched with a send.
+    pub interaction_id: T,
+    /// The tuple sent to the bus, with a selector.
+    /// The selector is an expression that might or might not reference
+    /// the multiplicity column, but it should always evaluate to a
+    /// binary value. If it is zero, the multiplicity must be zero as well.
+    /// For example, it could also be binary fixed column, indicating
+    /// where the multiplicity can be non-zero.
+    pub selected_tuple: SelectedExpressions<T>,
+    /// The multiplicity with which the tuple is sent to the bus.
+    /// None if this interaction comes from the RHS of a native lookup,
+    /// as native lookup arguments don't require the multiplicity to be
+    /// committed.
+    /// Note that this is the absolute multiplicity, i.e., for bus receives,
+    /// the negation has been removed.
+    pub multiplicity: Option<AlgebraicExpression<T>>,
+}
+
+impl<T: FieldElement> BusSend<T> {
+    /// Tries to find the matching receive, which should succeed iff. the send has
+    /// a static interaction ID.
+    pub fn try_match_static<'a>(
+        &self,
+        receives: &'a BTreeMap<T, BusReceive<T>>,
+    ) -> Option<&'a BusReceive<T>> {
+        Some(&receives[&self.interaction_id()?])
+    }
+
+    /// Returns the interaction ID if it is a number.
+    /// Sends and receives can be matched by matching the interaction ID.
+    pub fn interaction_id(&self) -> Option<T> {
+        match &self.interaction_id {
+            AlgebraicExpression::Number(id) => Some(*id),
+            _ => None,
+        }
+    }
+}
+
+impl<T: FieldElement> BusReceive<T> {
+    /// Returns whether the receive has an unconstrained multiplicity.
+    /// If it does not, this typically means that the multiplicity is binary constraint.
+    /// Unconstrained receives are equivalent to the RHS of a lookup;
+    /// binary-constrained receives are equivalent to the RHS of a permutation.
```

Comment:
```suggestion
    /// Returns true if the multiplicity of the receive can be an arbitrary value.
    /// This means that an arbitrary number of rows on the sending side with
    /// the same values can match a single row on the receiving end.
    /// Unconstrained receives are equivalent to the RHS of a lookup,
    /// while 0-1-constrained receives are equivalent to the RHS of a permutation.
```
(although, aren't they constrained to exactly 1?)

Reply:
> (although, aren't they constrained to exactly 1?)

what's exactly one? Many sends can map to the same receive. For example, for a range constraint lookup, you'd have one receive for all range constraints, with the multiplicities computed on all incoming sends.